plugins/main/tap/tap.go

```diff
@@ -0,0 +1,415 @@
+	"github.com/containernetworking/plugins/pkg/utils/sysctl"
+)
+
+const AddLink = "AddLink"
+
+type NetConf struct {
+	types.NetConf
+	MultiQueue bool   `json:"multiQueue"`
+	MTU        int    `json:"mtu"`
+	Mac        string `json:"mac,omitempty"`
+	Tapplugin  string `json:"tapplugin"`
+	Args       struct {
+		Mac string `json:"mac,omitempty"`
+	} `json:"args,omitempty"`
+	RuntimeConfig struct {
+		Mac string `json:"mac,omitempty"`
+	} `json:"runtimeConfig,omitempty"`
+}
+
+// MacEnvArgs represents CNI_ARG
+type MacEnvArgs struct {
+	types.CommonArgs
+	MAC types.UnmarshallableString `json:"mac,omitempty"`
+}
+
+func init() {
+	// this ensures that main runs only on main thread (thread group leader).
+	// since namespace ops (unshare, setns) are done for a single thread, we
+	// must ensure that the goroutine does not jump from OS thread to thread
+	runtime.LockOSThread()
+}
+
+func loadConf(args *skel.CmdArgs) (*NetConf, string, error) {
+	n := &NetConf{}
+	if err := json.Unmarshal(args.StdinData, n); err != nil {
+		return nil, "", fmt.Errorf("failed to load netconf: %v", err)
+	}
+	if args.Args != "" {
+		e := MacEnvArgs{}
+		err := types.LoadArgs(args.Args, &e)
+		if err != nil {
+			return nil, "", err
+		}
+
+		if e.MAC != "" {
+			n.Mac = string(e.MAC)
+		}
+	}
+
+	if n.Args.Mac != "" {
+		n.Mac = n.Args.Mac
+	}
+
+	if n.RuntimeConfig.Mac != "" {
+		n.Mac = n.RuntimeConfig.Mac
+	}
+
+	return n, n.CNIVersion, nil
+}
+
+func createLink(tmpName string, mtu int, nsFd int, multique bool, mac string) error {
+	linkAttrs := netlink.LinkAttrs{
+		Name:      tmpName,
+		Namespace: netlink.NsFd(nsFd),
+	}
+	if mtu != 0 {
+		linkAttrs.MTU = mtu
+	}
+	mv := &netlink.Tuntap{
+		LinkAttrs: linkAttrs,
+		Mode:      netlink.TUNTAP_MODE_TAP,
+	}
+	if mac != "" {
+		addr, err := net.ParseMAC(mac)
+		if err != nil {
+			return fmt.Errorf("invalid args %v for MAC addr: %v", mac, err)
+		}
+		linkAttrs.HardwareAddr = addr
+	}
+	if multique {
+		mv.Flags = netlink.TUNTAP_MULTI_QUEUE_DEFAULTS | netlink.TUNTAP_VNET_HDR | unix.IFF_TAP
+	}
+	if err := netlink.LinkAdd(mv); err != nil {
+		return fmt.Errorf("failed to create tap: %v", err)
+	}
+	return nil
+}
+
+func createTap(conf *NetConf, ifName string, netns ns.NetNS) (*current.Interface, error) {
+	tap := &current.Interface{}
+
+	// due to kernel bug we have to create with tmpName or it might
+	// collide with the name on the host and error out
+	tmpName, err := ip.RandomVethName()
+	if err != nil {
+		return nil, err
+	}
+
+	err = netns.Do(func(_ ns.NetNS) error {
+		if conf.Tapplugin != "" {
+			plugin, err := plugin.Open(conf.Tapplugin)
+			if err != nil {
+				return err
+			}
+			f, err := plugin.Lookup("CreateLink")
+			if err != nil {
+				return err
+			}
+			err = f.(func(tmpName string, mtu int, nsFd int, nsPath string, multique bool, mac string)(error))(tmpName, conf.MTU, int(netns.Fd()), netns.Path(), conf.MultiQueue, conf.Mac)
+			if err != nil {
+				return err
+			}
```

Comment:
This code does not check that 'CreateLink' has right function type (i.e. argument type and name). How about to add check the lookuped symbol has right function type as following sample code?

https://github.com/vladimirvivien/go-plugin-example/blob/070e14c627a7724f7ebe5c23f672ccb9181d39be/greeter.go#L40-L58

Reply:
fixed